File: frontend/src/types/Quote.ts
import { DiscountType } from '../constants/Enumeration/CommonEnum/DiscountEnum';
import { QuoteStatus } from '../constants/Enumeration/QuoteEnum/QuoteEnum';
import { TActivityHistory, TAddActivityHistory } from './ActivityHistory';
import { TCustomer } from './Customer';
import { TLineItem, TAddLineItem, TUpdateLineItem } from './LineItem';
import { TAddNote, TNote } from './Note';
import { TProperty } from './Property';
import { TUser } from './User';

export type TQuote = {
  id: string;
  workspaceId: string;
  jobId?: string;
  customerId: string;
  customer?: TCustomer;
  createdByUserId: string;
  createdByUser?: TUser;
  assignedToUserId: string;
  assignedToUser: TUser;
  title: string;
  propertyId: string;
  property?: TProperty;
  quoteNumber: string;
  status: QuoteStatus;
  sentAt?: string;
  viewed: boolean;
  viewedAt?: string;
  expiresAt: string;
  lineItems: TLineItem[];
  discountType: DiscountType;
  discountValue: number;
  taxRate: number;
  subtotal: number;
  discount: number;
  taxAmount: number;
  total: number;
  customerNotes: TNote[];
  internalNotes: TNote[];
  customerMessages?: string[];
  activityHistory: TActivityHistory[];
  source?: string;
  attachments: TQuoteAttachment[];
  createdAt: string;
  updatedAt: string;
};

export type TAddQuote = {
  workspaceId: string;
  customerId: string;
  createdByUserId: string;
  assignedToUserId: string;
  status: QuoteStatus;
  title: string;
  propertyId?: string;
  lineItems: TAddLineItem[];
  discountType: DiscountType;
  discountValue: number;
  taxRate: number;
  customerNotes?: TAddNote[];
  internalNotes?: TAddNote[];
  activityHistory: TAddActivityHistory[];
  source?: string;
};

export type TUpdateQuote = {
  id: string;
  assignedToUserId: string;
  lineItems: TUpdateLineItem[];
  title: string;
  propertyId: string;
  discountType: DiscountType;
  discountValue: number;
  taxRate: number;
  source: string;
};

export type TQuoteFilter = {
  q?: string;
  sortBy?: string;
  sort?: string;
  status?: string;
  createdMin?: string;
  createdMax?: string;
  totalMin?: number;
  totalMax?: number;
};

export type TQuoteAttachment = {
  id: string;
  fileName: string;
  url: string;
  quoteId: string;
  quote?: TQuote;
  createdAt: string;
};

export type TAddQuoteAttachment = {
  fileName: string;
  url: string;
};

export type TQuoteStats = {
  totalQuotes: number;
  totalValue: number;
  approvedValue: number;
  conversionRate: number;
};

File: backend/Models/QuoteModels/Quote.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
using backend.Services.Billing;

namespace backend.Models.QuoteModels;

public class Quote
{
    [Key]
    public Guid Id { get; set; }
    public Guid WorkspaceId { get; set; }
    public Guid? JobId { get; set; }

    public Guid CustomerId { get; set; }
    public Customer? Customer { get; set; }

    public Guid CreatedByUserId { get; set; }
    public User? CreatedByUser { get; set; }

    public Guid? AssignedToUserId { get; set; }
    public User? AssignedToUser { get; set; }

    public string Title { get; set; } = string.Empty;

    public Guid? PropertyId { get; set; }
    public Property? Property { get; set; }

    public string QuoteNumber { get; set; } = string.Empty;
    public QuoteStatus Status { get; set; } = QuoteStatus.Draft;

    public DateTime? SentAt { get; set; }
    public bool Viewed { get; set; } = false;
    public DateTime? ViewedAt { get; set; }
    public DateTime? ExpiresAt { get; set; }
    public string PaymentTerms { get; set; } = "uponReceipt";
    public decimal DepositAmount { get; set; }

    public ICollection<LineItem> LineItems { get; set; } = [];

    public DiscountType DiscountType { get; set; } = DiscountType.FixedAmount;
    public decimal DiscountValue { get; set; }
    public decimal TaxRate { get; set; }

    private TotalsBreakdown Totals => TotalsCalculator.Calculate(LineItems, DiscountType, DiscountValue, TaxRate);

    [NotMapped]
    public decimal Subtotal => Totals.Subtotal;

    [NotMapped]
    public decimal Discount => Totals.Discount;

    [NotMapped]
    public decimal TaxAmount => Totals.TaxAmount;

    [NotMapped]
    public decimal Total => Totals.Total;

    public List<Note> CustomerNotes { get; set; } = [];
    public List<Note> InternalNotes { get; set; } = [];

    public List<string> CustomerMessages { get; set; } = [];
    public List<ActivityHistory> ActivityHistory { get; set; } = [];

    public string Source { get; set; } = string.Empty;

    public ICollection<QuoteAttachment> Attachments { get; set; } = [];

    public bool IsArchived { get; set; } = false;
    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;
}

public enum QuoteStatus
{
    Draft,
    Sent,
    AwaitingResponse,
    AwaitingApproval,
    Approved,
    Declined,
    Expired,
    ConvertedToJob
}

public enum DiscountType
{
    Percentage,
    FixedAmount
}

public enum QuoteActivityType
{
    QuoteCreated,
    QuoteEdited,
    QuoteSent,
    InternalNoteAdded,
    CustomerNoteAdded,
    CustomerMessageAdded,
    AttachmentAdded,
    MarkedSent,
    MarkedAccepted,
    MarkedRejected,
    ConvertedToJob,
    StatusChanged
}
